Add the ability to edit options settings for providers

enoodle opened the pull request:
Per provider options were added in [1][2] to all providers. It is now possible to save values in `ems.options` to use as options.
This PR adds the ability to set options from the UI. To tell which options are supported for each provider type an API request is done [3] which expects the provider to implement the `provider_settings` method, for example [3],[4].
It will look for two sections in the returned value "proxy_settings" and "advanced_settings" and will display them in their own tab. for each section there are expected to be a "label" and "help_text" fields and a "settings" field with all of its options (or sub sections). The options themselves should also have "label" and "help_text" fields and also a "global_default" field that may exist.

![provider_options_new_provider](https://user-images.githubusercontent.com/3123328/31306404-4f2d6590-ab58-11e7-9462-4c51e49664a4.png)

based on:
[1] https://github.com/ManageIQ/manageiq/pull/15398 - merged
[2] https://github.com/ManageIQ/manageiq-schema/pull/23 - merged
[3] https://github.com/ManageIQ/manageiq-providers-kubernetes/pull/45 - merged
[4] https://github.com/ManageIQ/manageiq-providers-openshift/pull/32 - merged
[5] https://github.com/ManageIQ/manageiq/pull/15799 - merged
[6] https://github.com/ManageIQ/manageiq-api/pull/3 - merged

This aims to provide solution to a few RFEs:
https://bugzilla.redhat.com/show_bug.cgi?id=1379185
https://bugzilla.redhat.com/show_bug.cgi?id=1459189 (might be a duplicate of the previous one?)
https://bugzilla.redhat.com/show_bug.cgi?id=1459555
https://bugzilla.redhat.com/show_bug.cgi?id=1462835

enoodle commented:
I am now having a problem with creating a new provider: the ems type that is chosen is the basic `ManageIQ::Providers::ContainerManager` which doesn't have any options. When the user chooses a type, I want to update the options, but all the pages are already generated. Is it possible to refresh the page with the chosen container manager type? Is it even the right solution?
@martinpovolny @himdel @mzazrivec

himdel commented:
@enoodle The provider edit screen is always an angular screen now... So we can't easily re-render any HTML that gets rendered when opening the form.

So I think you'll need to move the logic to angular, query a json endpoint to get those options after the provider type is selected, and do all the option-dependent rendering in angular instead of server-side.

(A workaround would be to create a component that can recieve HTML from the server and render it, and route it through there, but.. I think it's easier to do it properly on the js side :).)

enoodle commented:
@himdel This is now rendering the HTML for the options forms in angular but I am sure that there is a much better way to do it. Can you take a look?

himdel commented on app/assets/javascripts/controllers/ems_common/ems_common_form_controller.js:
```diff
@@ -557,5 +563,46 @@ ManageIQ.angular.app.controller('emsCommonFormController', ['$http', '$scope', '
+      $scope.emsCommonModel.provider_options_description = response;
+      if ($scope.emsCommonModel.provider_options_description.hasOwnProperty('advanced_settings')) {
+        var advanced_settings = $scope.emsCommonModel.provider_options_description.advanced_settings.settings
+        for (section in advanced_settings) {
```
Careful about creating global variables here ;) `var section`

himdel commented on app/assets/javascripts/controllers/ems_common/ems_common_form_controller.js:
```diff
@@ -557,5 +563,46 @@ ManageIQ.angular.app.controller('emsCommonFormController', ['$http', '$scope', '
+    settings_group[section_name] = options_section;
+    settings_group[section_name].section_name = section_name;
+    var section = settings_group[section_name].settings;
+    for (option in section){
```
Another global ;)

himdel commented on app/assets/javascripts/controllers/ems_common/ems_common_form_controller.js:
```diff
@@ -557,5 +563,46 @@ ManageIQ.angular.app.controller('emsCommonFormController', ['$http', '$scope', '
+    }
+  };
+
+  $scope.toSnakeCase = function(st) {
```
You can probably use [_.snakeCase()](https://lodash.com/docs/3.10.1#snakeCase) instead..

himdel commented on app/assets/javascripts/controllers/ems_common/ems_common_form_controller.js:
```diff
@@ -11,6 +11,8 @@ ManageIQ.angular.app.controller('emsCommonFormController', ['$http', '$scope', '
       default_hostname: '',
       amqp_hostname: '',
       hawkular_hostname: '',
+      provider_options_values: {},
```
Not quite sure what these are, but looks like they are generic options, not related to the specific instance you're editing, is that so?

If so, you probably want `$scope.somethingElse` :) - the idea is that the `model` object should be just the data of that particular instance (not necessarily quite 1:1 with the database, but as close as possible, so that you can just post the whole thing to the server, and not have to delete fields when saving.).

enoodle replied:
I kind of need the same object to hold both the descriptions of the options and the values. This way it is much easier to iterate over the descriptions in the haml file and find the object that should hold the value.
Another option can be to hold it outside of emsCommonModel and only copy the value of the options back to it when the user clicks "save". WDYT?

himdel replied:
Ah, got it :).

Well, it will probably work just fine as it is, I'm a bit reluctant to try to force you to make this super-clean when that whole screen is super-dirty already :).

That said, yes I would work that other approach better. If the structure is relatively flat, you could also `$watch` it for changes and update the model value on any change.

himdel commented on app/views/layouts/angular-bootstrap/_provider_advanced_settings.html.haml:
```diff
@@ -0,0 +1,27 @@
+  %p
+    {{ settings_section.help_text }}
+  %div{"ng-repeat" => "option in settings_section.settings"}
+    %div{"ng-show" => "true"}
```
ng-show true .. Probably can get rid of it?

himdel commented on app/views/layouts/angular-bootstrap/_provider_proxy_settings.html.haml:
```diff
@@ -0,0 +1,24 @@
+%p
+  {{ emsCommonModel.provider_options_description.proxy_settings.help_text }}
+%div{"ng-repeat" => "option in emsCommonModel.proxy_settings.settings"}
+  %div{"ng-show" => "true"}
```
same here..

himdel commented on app/views/layouts/angular-bootstrap/_provider_advanced_settings.html.haml:
```diff
@@ -0,0 +1,27 @@
+                              "placeholder"   => "{{ option.global_default }}",
+                              "title"         => "{{ option.help_text }}",
+                              "prefix"        => "provider_options"}
+          %span.help-block{"ng-attr-ng-show" => "{{ \"angularForm.provider_options_\" + settings_section.section_name + \"_\" + option.name + \".$error.detectedSpaces\" }}"}
```
You should never need to do this... (`ng-attr-` with another angular directive)

`"ng-show" => 'angularForm["provider_options_" + settings_section.section_name + "_" + option.name].$error.detectedSpaces'` is probably what you want..

himdel replied:
(We should really start putting these as sub-objects, and drop the idea of using prefixed fields, but .. I guess that's too big a change for the PR.)

himdel commented on app/views/layouts/angular-bootstrap/_provider_advanced_settings.html.haml:
```diff
@@ -0,0 +1,27 @@
+    {{ settings_section.help_text }}
+  %div{"ng-repeat" => "option in settings_section.settings"}
+    %div{"ng-show" => "true"}
+      .form-group{"ng-attr-ng-class" => "{{ \"{'has_error': angularForm.provider_options_\" + settings_section.section_name + \"_\" + option.name + \".$invalid}\" }}"}
```
`"ng-class" => '{ "has_error": angularForm["provider_options_" + settings_section.section_name + "_" + option.name].$invalid'`

himdel commented:
@enoodle Looks like you're on the right track..

Maybe since this is a new functionality, you may want to wrap this in a component, but not sure if this is separate enough from the rest to make that practical..

We actually already have a component which takes a definition object with a list of fields, and renders those fields, but not sure if this is close enough so that you could actually reuse it, or if it's more like a place to take inspiration from.. https://github.com/ManageIQ/manageiq-ui-classic/blob/master/app/assets/javascripts/components/ansible-credential-options.js


But I have no better idea than what you're already doing, so the component thing is mostly about furure tech debt and minimizing the dependency on shared angularForm, etc., it probably wouldn't make anything easier for you..

enoodle commented:
@himdel as you can see I am using two pretty similar "haml" files here. one of them, the _provider_proxy_settings.html.haml, is kind of included in the other one.
Is there a way that I can re-use this logic but still use different angular objects?

himdel commented:
@enoodle Re https://github.com/ManageIQ/manageiq-ui-classic/pull/1652#issuecomment-321210081 .. well it looks like the template is mostly the same, but the field names are completely different.

So I guess there's 2 ways to re-use..

Either on the partials level - you can `= render :partial => "path"` one from the other, and pass the differences in `:locals`. That would be easy to do, but probably not so easy to un-tangle afterwards.

Or you can go for the component, passing all these bit explicitly to the component, so the differences are limited to how you *use* the component and not really inside it.

---

Do you want to do something like this or should we just merge?

enoodle commented:
@miq-bot add_label wip
@himdel I don't think it is ready for merge yet. I now saw that it is possible to write specs for this too and I think it would be better to do.

enoodle commented:
@himdel I have a new problem. I suddenly started getting these errors on the input fields when I write into them. Do you happen to know what changed?

```
TypeError: Cannot read property '$setDirty' of undefined
    at viewModelComparison (checkchange.self-ead9614….js?body=1:59)
    at checkchange.self-ead9614….js?body=1:28
    at Scope.$digest (angular.self-6b939a3….js?body=1:18211)
    at angular.self-6b939a3….js?body=1:18412
    at completeOutstandingRequest (angular.self-6b939a3….js?body=1:6363)
    at angular.self-6b939a3….js?body=1:6643
(anonymous) @ angular.self-6b939a3….js?body=1:14643
(anonymous) @ angular.self-6b939a3….js?body=1:11103
$digest @ angular.self-6b939a3….js?body=1:18229
(anonymous) @ angular.self-6b939a3….js?body=1:18412
completeOutstandingRequest @ angular.self-6b939a3….js?body=1:6363
(anonymous) @ angular.self-6b939a3….js?body=1:6643
angular.self-6b939a3….js?body=1:14643 TypeError: Cannot read property '$pristine' of undefined
    at Array.<anonymous> (checkchange.self-ead9614….js?body=1:41)
    at Object.$$parseAndValidate (angular.self-6b939a3….js?body=1:29061)
    at Object.$commitViewValue (angular.self-6b939a3….js?body=1:29049)
    at angular.self-6b939a3….js?body=1:29190
    at ChildScope.$eval (angular.self-6b939a3….js?body=1:18373)
    at ChildScope.$apply (angular.self-6b939a3….js?body=1:18473)
    at Object.$$debounceViewValueCommit (angular.self-6b939a3….js?body=1:29189)
    at Object.$setViewValue (angular.self-6b939a3….js?body=1:29167)
    at HTMLInputElement.listener (angular.self-6b939a3….js?body=1:25077)
    at HTMLInputElement.dispatch (jquery.self-25ca496….js?body=1:4738)
(anonymous) @ angular.self-6b939a3….js?body=1:14643
(anonymous) @ angular.self-6b939a3….js?body=1:11103
$apply @ angular.self-6b939a3….js?body=1:18478
$$debounceViewValueCommit @ angular.self-6b939a3….js?body=1:29189
$setViewValue @ angular.self-6b939a3….js?body=1:29167
listener @ angular.self-6b939a3….js?body=1:25077
dispatch @ jquery.self-25ca496….js?body=1:4738
elemData.handle @ jquery.self-25ca496….js?body=1:4550
```

himdel commented:
@enoodle That looks like you have a `checkchange` (and thus presumably an input) outside the form. Either really really outside, or in a isolated scope, so it can't see `angularForm` in the current scope.

But unless you're using components (or a directive with `scope: {}`), you wouln't have an isolate scope, so.. probably outside the form?

enoodle commented:
@himdel 
It is indeed from the `checkchange` attribute, but I don't understand how it is outside of a form. when I look at the dom of the page I clearly see it inside the form that is created in "app/views/ems_container/new.html.haml".  I am pretty sure that this used to work before and stopped but I am not sure what caused it.

If you want to test it you need to apply the manageiq-api patch ManageIQ/manageiq-api#3 and also https://github.com/ManageIQ/manageiq-providers-kubernetes/pull/91 because of name changing requirements. No need for a real provider as this is shown in the new provider creation screen.

himdel commented:
@enoodle If you want to experiment with this before I can test, when you inspect the `input` DOM element in the browser, you can then do `scope = angular.element($0).scope()` to see the scope the input lives in.

Then, `scope.angularForm` should return *something*, if it does, then it should work :).
If it doesn't, try `scope.$parent.angularForm` or `scope.$parent.$parent.angularForm`, etc..

If it's not there at all, then it really is somehow outside that form.

If it is there but you need a bunch of `$parent`, then you know some directive between the form and the input is breaking this by introducing isolate scope. 

Nested scopes use prototypal inheritance by default, so `scope.foo` should always work when `scope.$parent.foo` does. But when you have an isolate scope, the parent is only available explicitly via `$parent` and the prototype goes to `{}`. Which means the topmost scope where `scope.$parent.angularForm` works but `scope.angularForm` doesn't is the culprit :).

(Usually the fastest way to find which element a particular scope corresponds to is just to count them and then go up the DOM and look at classes - only elements with a `ng-scope` class introduce a new scope.)

enoodle commented:
@himdel 
`angular.element("#provider_options_image_inspector_options_http_proxy")[0]` - This returns the input field, but it doesn't have a `scope` or `$parent` method/attribute. its `form` attribute seems to be the correct form. It third `parentElement` has the ng-scope class, it is the `ng-repeat` tag from that iterates over the options [1] so I can't change it (I didn't specify in the haml anything about its class). Any advice?


[1] https://github.com/ManageIQ/manageiq-ui-classic/pull/1652/files#diff-ff9cae74cde297ab63b638b11e9b7e14R10

himdel commented:
Well, `angular.element("#provider_options_image_inspector_options_http_proxy")[0]` doesn't have `scope`, but `angular.element("#provider_options_image_inspector_options_http_proxy")` does ;)

enoodle commented:
Tests are green now, but the errors listed above still appear in the console in the browser

@AparnaKarve Can you please take a look? I tried to debug it with @himdel  but we couldn't find the source of it. Now that all the dependencies have been merged maybe you can try it.

AparnaKarve commented:
@enoodle I pulled down your PR but did not see the 2 new tabs appear, after I chose Kubernetes.
For the tabs to appear, I even removed the 2 `ng-if`s below, but no luck.

```
'ng-if' => "emsOptionsModel.provider_options.hasOwnProperty('proxy_settings')"}) 
```
and
```
'ng-if' => "emsOptionsModel.provider_options.hasOwnProperty('advanced_settings')"})
```

Can you check too?
To be able to see your `checkchange` related issue, I would need the tabs to appear. Thanks.

enoodle commented:
@AparnaKarve Thanks for taking a look at this. I think maybe  you didn't have the `manageiq-providers-kubernetes`  repo updated? I guess you are updating the others more regularly, but for this one the gem might not be updated because we didn't make any release of it yet. (The tabs to appear for me)

himdel commented:
I can see the error now, it looks like a combination of using computed `name`, `ng-repeat` and `checkchange`.

Checkchange needs every input to have a name, and it needs that name not to change...

You're getting

```
VM6722:27 TypeError: Cannot read property '$setDirty' of undefined
```

on

```js
scope.angularForm[scope['formchange_' + ctrl.$name]].$setDirty();
```

because `ctrl.$name` is `provider_options_proxy_settings_http_proxy`, but `scope['formchange_' + ctrl.$name]` evaluates to `"{{ "provider_options_proxy_settings_" + option.name }}"` .. which is not quite available under `angularForm` :).

Why that evaulates to that is that `checkchange` does

```js
scope['formchange_' + ctrl.$name] = elem[0].name;
```

in the `link` function, so it needs the name to be right then.

enoodle commented:
> So.. as I recommended earlier, make the inside of those tabs a component with explicit inputs and outputs, you'll save both yourself and us a lot of time :) (beacuse if you don't, somebody will have to do it soon either way :)).

IIUC even if I will build a component for the tab content I will have to explicitly define the input fields in it and won't be able to create them dynamically from the provider options? I would like to find a way to build those tabs dynamically because maintaining it will be a trouble and should be avoided.

Where is the `checkchange` function defined? I would like to see how I can work with it to use input fields created with `ng-repeat`.

himdel commented:
@enoodle My point was .. the field being called `provider_options_proxy_settings` is an implementation detail, and this whole "creating names by concatenating strings" idea is simply broken from the start.

So, eventually..

* we want each tab to be a separate form and component, (angular supports nested forms quite well)
* any data passing between the main form and this tabs needs to be made explicit - not changing the huge model object right away
* that way, your component can deal with predictable names that don't depend on anything from the outside - most notably, they should not depend on the tab name
* and it will be the parent component's responsibility do put the data in the right place once it receives it.
* also all those prefixes should become subobjects (that will probably have to come last) - no more `default_userid` and `default_password` when you can have `default: {userid, password}` and not care which tab you're in

If your inputs actually depend on the data, this still helps you because by isolating into a component and making all that "are you ok?" and "are you validated?" explicit, you no longer need to be using `checkchange` or anything that's slowing you down, as long as you take care of doing the equivalent functionality your way.

FYI we're replacing `checkchange` (which is per-input and creates custom stuff on scope) with `form-changed` (which is per-form and just compares model to modelCopy) in the rest of the code, but EmsCommon is a tough one :).

As for where it's defined.. `app/assets/javascripts/directives/checkchange.js`

enoodle commented:
@himdel I made something crud and I need help to cover your points. 

> any data passing between the main form and this tabs needs to be made explicit - not changing the huge model object right away but calling a function to do that

where those calls would be made? how can the form controller get hold of the component to call its functions?
Should I add a validation function to `providerOptionFieldInput` that will be called from `providerOptionSection` that will be called from `emsCommonFormController` when in `isBasicInfoValid` ? How would that last controller get the components?

AparnaKarve commented:
@enoodle Are you adopting the component approach for those 2 tabs in order to avoid dealing with the `checkchange` error?
Please note that the EMS form is a relatively complex form, and therefore if possible, you should hold off on making major changes until the form has been converted to either a `component` or at least to `controllerAs`

Which field specifically was giving you the `checkchange` related error?
If it's a field in your newly added tab, then for the time being, you can even consider not using the `checkchange` directive on that field. 
Not using `checkchange` would mean no dirty-checking on that field -- it's a trade-off, yes...but it would be better than adding any risky changes.

Once the form has been officially converted to a component, all the fields would be automatically accounted for dirty-checking.

enoodle commented:
@himdel @AparnaKarve I removed the "checkchange" attribute from the input fields. It works without errors now, probably not being checked for bad inputs too much but I guess it is ok for now.
Do you want me to leave it with the simple components that I have created? I can remove them easily as they are only in the second commit here.

dclarizio commented:
@enoodle are `Proxy Settings` and `Advanced Settings` really Endpoints, as seen in the screen shot?  I'm worried we're overloading those tabs which should only contain actual Endpoints.

@simon3z can you weigh in on this design (is there a design documented somewhere)?  Perhaps these options should not be part of the Endpoints area on the screen.

AparnaKarve commented:
> are Proxy Settings and Advanced Settings really Endpoints, as seen in the screen shot?

@dclarizio That's a good point.

@enoodle Endpoints are associated with the Validation of the provider.
Do ` Proxy Settings` and `Advanced Settings` just seem like regular (and optional) fields  for Kubernetes? If so, they should be placed on top in the main form itself and not in the credentials/endpoints area.
You could perhaps use Radio buttons for ` Proxy Settings` and `Advanced Settings`, and depending on the user selection display the corresponding fields.

simon3z commented:
> @simon3z can you weigh in on this design (is there a design documented somewhere)?

@dclarizio UX was designed with @loicavenel and @serenamarie125 
The mocks were posted for comments in an internal thread/ml and presented at the all-hands. Sharing one of the relevant mocks here for reference:

![editproviderchanges](https://user-images.githubusercontent.com/130393/30984467-a84f6794-a48d-11e7-8fb1-3780eebfa984.png)

> Perhaps these options should not be part of the Endpoints area on the screen.

They're not because the "Endpoint" title has been removed (also mentioned on the thread, I'll provide references in private).

I feel that this PR has been dragged in way too many directions and rewritten in several possible ways.
In general that's not an argument for quality but there's also a limit to the number of reevaluations and rewrites that we can do.

If you feel that we need another meeting to get to the bottom of it, let's do it and we'll see what we can get to do to improve this if needed.

@miq-bot assign himdel

dclarizio commented:
@simon3z I apologize, as I didn't get to see that presentation at the all hands.
> They're not because the "Endpoint" title has been removed (also mentioned on the thread, I'll provide references in private).

The reason I mentioned this is because the GIF in the top comment of this PR still has that title in it.

@Loicavenel / @serenamarie125 Removing the `Endpoints` title leaves all of these tabs out of context.  The tabs used to mean Default Endpoint, Metrics Endpoint, etc.  Now it's Default what?  Metrics what? Advanced what? etc . . . 

Can we put some kind of title to help explain what the tabs are for?  Maybe Credential Information or Security Options?

simon3z commented:
> @simon3z I apologize, as I didn't get to see that presentation at the all hands.

@dclarizio NP, too many interesting tracks :blush: 

@dclarizio @Loicavenel @serenamarie125 do you feel that it would be more efficient to meet to discuss the questions/options above?

enoodle commented:
@dclarizio @AparnaKarve I have updated the top comment with the output on the master branch. The "Endpoints" title is already removed.

dclarizio commented:
Thanks @enoodle, we plan to discuss this early next week as to if we need a title or not above those tabs to give some context.
 . . . Dan

simon3z commented:
> Thanks @enoodle, we plan to discuss this early next week as to if we need a title or not above those tabs to give some context.

@dclarizio @enoodle @Loicavenel @serenamarie125 I sent an invite

enoodle commented:
@dclarizio @AparnaKarve Where should I put the titles "Endpoints" and "Settings" ? What type of element they should be in?

AparnaKarve commented:
@enoodle I did a bit of digging around in github, and found the PR and commit that removed the `Endpoints` title.

PR - https://github.com/ManageIQ/manageiq-ui-classic/pull/1501
commit - https://github.com/ManageIQ/manageiq-ui-classic/pull/1501/commits/39f61a073a768e012cc33af6ba79de027cfdbba7

Ideally, removing the `Endpoints` title should have been discussed in that PR. But, anyway...

So, all you need to do is add the `Endpoints` title back, (i.e. undo the `Endpoints` title related changes in  https://github.com/ManageIQ/manageiq-ui-classic/pull/1501/commits/39f61a073a768e012cc33af6ba79de027cfdbba7)
And along those similar lines, add the `Settings` title for the new `Settings` tabbed section.

AparnaKarve commented on app/views/layouts/angular/_per_provider_settings.html.haml:
```diff
@@ -0,0 +1,25 @@
+#settings_tabs
+  %h3
+    Settings
```
Change this to `_("Settings")` for i18n

enoodle commented:
@AparnaKarve I made those changes. Updated the top comment too.

AparnaKarve commented:
@enoodle Thanks. Changes look in line with what we discussed in our call.

I'm wondering if we should move the text in Proxy/Advanced settings that you are currently displaying above the input fields (on the left), below the input fields - just so that it looks less busy.

simon3z commented:
@enoodle in the screenshot of the description there is a "Required. Used to gather utilization data" which shouldn't be there for the "Default" endpoint I guess.

enoodle commented:
@simon3z I am seeing this line there on master branch as well. Maybe it was removed since I last rebased but anyway it is not connected to the new settings section.

enoodle commented:
> I'm wondering if we should move the text in Proxy/Advanced settings that you are currently displaying above the input fields (on the left), below the input fields - just so that it looks less busy.

@AparnaKarve I am sorry I am not following. Do you mean the descriptions of the settings:
```
Advanced Settings

Advanced Settings for provider configuration

Image Inspector Options

Settings for Image Inspector tool
```
I think it should come before the options, but maybe it should be changed in a different PR that will fine tune the appearance of these tabs here?

AparnaKarve commented:
@enoodle Yes, that is correct. I was referring to all that text (the settings descriptions) that you display right at the top of the Proxy/Advanced tabs.
First off, do we need those descriptions? Aren't the labels self-explanatory?

But if you do need to keep those descriptions, IMO, if you could move it down, it would just be more easy on the eyes (less stuff on the screen or less busy)

enoodle commented:
@AparnaKarve I removed these lines. PTAL

AparnaKarve commented:
@enoodle Looks good now. Thanks.

I did notice something with the `Validate` button.

When I select Kubernetes and fill in all the fields under the `Default` tab, the Validate button is enabled.

Now, when I merely click on the `Advanced Settings` tab, the Validate button is disabled.
Shouldn't it stay enabled?

AparnaKarve commented:
One more thing -

Since you already have a `Settings` title, maybe the tabs should just be called -
`Proxy` and `Advanced` ( I think you can drop the word `Settings` from the tabs)

enoodle commented:
@AparnaKarve I removed the "Settings" from the tabs titles and added the tabs name in `validateBasicInfo()` to fix the problem with the validate button. PTAL

enoodle commented:
@AparnaKarve @himdel I fixed the CodeClimate error, PTAL

ohadlevy commented on app/views/layouts/angular/_per_provider_settings.html.haml:
```diff
@@ -0,0 +1,25 @@
+  %ul.nav.nav-tabs
+    = miq_tab_header('proxy_settings', nil, {'ng-click' => "changeAuthTab('proxy_settings')",
+                                             'ng-if'    => "emsOptionsModel.provider_options.hasOwnProperty('proxy_settings')"}) do
+      %i{"error-on-tab" => "hawkular", :style => "color:#cc0000"}
```
just wondering, is this file suppose to be generic? if thats the case, why do you have ```hawkular``` here?

enoodle replied:
Thanks, It was a mistake, I changed it to "proxy_settings" and "advanced_settings" below.

enoodle commented:
ping @AparnaKarve @himdel @ohadlevy can you take another look?

AparnaKarve commented:
@enoodle Thank you for your patience.
I have been tied up with other deliverables required for the release and hence was not able to review this sooner. 
I checked out your latest update and the above issue seems resolved, which is good.

I did see something with the way the tabs are positioned when `Prometheus` is selected from the `Alerts` dropdown, but it is purely an aesthetic issue and may not be even related to the code changes in this PR.

Notice the gap between the `Default` and the `Alerts` tab. That gap gets filled up when I select a value from the `Metrics` dropdown. 
<img width="746" alt="screen shot 2017-10-16 at 9 21 05 am" src="https://user-images.githubusercontent.com/1538216/31623367-0a8911b6-b254-11e7-8ac1-9283419fdada.png">




I think we can treat that as a separate (and unrelated) issue and get this PR merged.

@dclarizio IMO, this is GTG. Thanks.

dclarizio commented:
@enoodle There's a lot of new code here, could you add some tests, at least basic ones?  @AparnaKarve can you suggest some tests that can be added?

AparnaKarve commented:
@enoodle On the backend side, can you write a spec that will create a `Kubernetes` record taking into account all those newly added fields for `Proxy` and `Advanced`?

On the JS side, I see that you have included a few tests.
Could you also write a basic `GET` test for `Kubernetes`?(you can refer to the existing `$httpBackend.whenGET` tests) 

The `emsCommonFormController` angular controller would be undergoing a refactor in a couple of weeks from now. What would really help us at that point, are the JS specs that test the code for various use cases - the specs will be an assurance for us that we have not broken anything while refactoring.

I guess some of the basic use cases for this PR would be how the `emsCommonModel` gets populated based on what the user selects from the dropdown: `Prometheus`, `Hawkular` etc.
So a few JS specs along those lines would be nice. 
Thanks.

enoodle commented:
@AparnaKarve 

> can you write a spec that will create a Kubernetes record taking into account all those newly added fields for Proxy and Advanced?

I am not sure where this should be, can you direct me to the spec file?

> Could you also write a basic GET test for Kubernetes?(you can refer to the existing $httpBackend.whenGET tests)

I added a few tests in similar to other examples in the spec file, PTAL

> I guess some of the basic use cases for this PR would be how the emsCommonModel gets populated based on what the user selects from the dropdown: Prometheus, Hawkular etc.

This PR is for the Options only, I don't want to create Prometheus/Hawkular data that I am not sure about here. I added a test in `ems_common_controller_sepc.rb` that I hope covers it for the options. PTAL

AparnaKarve commented:
> I am not sure where this should be, can you direct me to the spec file?

You could add specs for `Kubernetes` here - spec/controllers/ems_container_controller_spec.rb

Check out some similar examples in spec/controllers/ems_infra_controller_spec.rb where there are specs written for Create/Update/Validate for `SCVMM`, `OpenStack`, `RHEV`, `VMWare`

enoodle commented:
@AparnaKarve I added a create test with provider options to spec/controllers/ems_container_controller_spec.rb 
PTAL

AparnaKarve commented:
@enoodle Thanks for the specs.

@dclarizio This can be merged now. Thanks.
